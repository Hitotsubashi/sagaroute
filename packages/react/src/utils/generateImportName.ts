import path from 'path';

function capitalize(word: string) {
  return word.charAt(0).toUpperCase() + word.slice(1);
}

export default function generateImportName(fpath: string, varName: string): string {
  const { name, dir } = path.parse(fpath);
  let upper = true;
  let importName = '';
  let basepath = path.join(dir, name);
  for (let i = 0; i < basepath.length; i++) {
    // if (basepath[i] === ".") {
    //   upper = false;
    // } else if ([path.sep, "-", "_", "[", "]", "$"].includes(basepath[i])) {
    if (!/[A-Za-z0-9]/.test(basepath[i])) {
      upper = true;
    } else {
      if (upper) {
        importName += basepath[i].toUpperCase();
        upper = false;
      } else {
        importName += basepath[i];
      }
    }
  }
  if (!/[_A-Za-z]/.test(importName[0])) {
    importName = '_' + importName;
  }
  if (name === '[]') {
    importName += '_';
  }
  return importName + capitalize(varName);
}
