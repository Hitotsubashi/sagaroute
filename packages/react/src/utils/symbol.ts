export const EVAL_STRING_SYMBOL = '/*#__EVAL__*/';
export const EVAL_STRING_SYMBOL_REGEXP = /"\/\*#__EVAL__\*\/(.*?)\/\*#__EVAL__\*\/"/g;
export const ROUTE_CONTENT_SYMBOL_REGEXP = /SAGA_ROUTE:?(.*)?/;
export const TEMPLATE_RENDER_ROUTE_SYMBOL = 'routes';
export const TEMPLATE_RENDER_IMPORTS_SYMBOL = 'imports';
export const INJECT_SYMBOL_PREFIX = 'sagaroute-inject';
export const INJECT_SYMBOL_PREFIX_REGEXP = /sagaroute-inject:(.*)/;

export const htmlTagNames = [
  'a',
  'abbr',
  'acronym',
  'address',
  'applet',
  'area',
  'article',
  'aside',
  'audio',
  'b',
  'base',
  'basefont',
  'bdi',
  'bdo',
  'bgsound',
  'big',
  'blink',
  'blockquote',
  'body',
  'br',
  'button',
  'canvas',
  'caption',
  'center',
  'cite',
  'code',
  'col',
  'colgroup',
  'command',
  'content',
  'data',
  'datalist',
  'dd',
  'del',
  'details',
  'dfn',
  'dialog',
  'dir',
  'div',
  'dl',
  'dt',
  'element',
  'em',
  'embed',
  'fieldset',
  'figcaption',
  'figure',
  'font',
  'footer',
  'form',
  'frame',
  'frameset',
  'h1',
  'h2',
  'h3',
  'h4',
  'h5',
  'h6',
  'head',
  'header',
  'hgroup',
  'hr',
  'html',
  'i',
  'iframe',
  'image',
  'img',
  'input',
  'ins',
  'isindex',
  'kbd',
  'keygen',
  'label',
  'legend',
  'li',
  'link',
  'listing',
  'main',
  'map',
  'mark',
  'marquee',
  'math',
  'menu',
  'menuitem',
  'meta',
  'meter',
  'multicol',
  'nav',
  'nextid',
  'nobr',
  'noembed',
  'noframes',
  'noscript',
  'object',
  'ol',
  'optgroup',
  'option',
  'output',
  'p',
  'param',
  'picture',
  'plaintext',
  'pre',
  'progress',
  'q',
  'rb',
  'rbc',
  'rp',
  'rt',
  'rtc',
  'ruby',
  's',
  'samp',
  'script',
  'search',
  'section',
  'select',
  'shadow',
  'slot',
  'small',
  'source',
  'spacer',
  'span',
  'strike',
  'strong',
  'style',
  'sub',
  'summary',
  'sup',
  'svg',
  'table',
  'tbody',
  'td',
  'template',
  'textarea',
  'tfoot',
  'th',
  'thead',
  'time',
  'title',
  'tr',
  'track',
  'tt',
  'u',
  'ul',
  'var',
  'video',
  'wbr',
  'xmp',
];

export const svgTagNames = [
  'a',
  'altGlyph',
  'altGlyphDef',
  'altGlyphItem',
  'animate',
  'animateColor',
  'animateMotion',
  'animateTransform',
  'animation',
  'audio',
  'canvas',
  'circle',
  'clipPath',
  'color-profile',
  'cursor',
  'defs',
  'desc',
  'discard',
  'ellipse',
  'feBlend',
  'feColorMatrix',
  'feComponentTransfer',
  'feComposite',
  'feConvolveMatrix',
  'feDiffuseLighting',
  'feDisplacementMap',
  'feDistantLight',
  'feDropShadow',
  'feFlood',
  'feFuncA',
  'feFuncB',
  'feFuncG',
  'feFuncR',
  'feGaussianBlur',
  'feImage',
  'feMerge',
  'feMergeNode',
  'feMorphology',
  'feOffset',
  'fePointLight',
  'feSpecularLighting',
  'feSpotLight',
  'feTile',
  'feTurbulence',
  'filter',
  'font',
  'font-face',
  'font-face-format',
  'font-face-name',
  'font-face-src',
  'font-face-uri',
  'foreignObject',
  'g',
  'glyph',
  'glyphRef',
  'handler',
  'hkern',
  'iframe',
  'image',
  'line',
  'linearGradient',
  'listener',
  'marker',
  'mask',
  'metadata',
  'missing-glyph',
  'mpath',
  'path',
  'pattern',
  'polygon',
  'polyline',
  'prefetch',
  'radialGradient',
  'rect',
  'script',
  'set',
  'solidColor',
  'stop',
  'style',
  'svg',
  'switch',
  'symbol',
  'tbreak',
  'text',
  'textArea',
  'textPath',
  'title',
  'tref',
  'tspan',
  'unknown',
  'use',
  'video',
  'view',
  'vkern',
];
