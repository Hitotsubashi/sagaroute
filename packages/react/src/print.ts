import path from 'path';
import fs from 'fs';
import {
  INJECT_SYMBOL_PREFIX_REGEXP,
  TEMPLATE_RENDER_IMPORTS_SYMBOL,
  TEMPLATE_RENDER_ROUTE_SYMBOL,
} from '@/utils/symbol';
import { Imports, RouteObject } from '@/weave';
import { ParseResult } from '@babel/parser';
import traverse from '@babel/traverse';
import * as t from '@babel/types';
import generate from '@babel/generator';
import Mustache from 'mustache';
import hookCompose from '@/utils/hookCompose';
import parseToAst from '@/utils/parseToAst';
import transformRoutesToString from './utils/transformRoutesToString';

export type PrintHookBeforeParse = (routeFilePath: string) => void | string | null;
export type PrintHookAfterParse = (template: string, routeFilePath: string) => void;
export type PrintHookBeforeInject = (
  view: Record<string, any>,
  template: string,
  routes: RouteObject[],
) => void | string | null;
export type PrintHookAfterInject = (
  content: string,
  view: Record<string, any>,
  template: string,
) => void;
export type PrintHookBeforeWrite = (renderedContent: string, routeFilePath: string) => void | null;
export type PrintHookAfterWrite = (routeFilePath: string) => void;

interface PrintOption {
  routeFilePath?: string;
  onWarning?: (message: string) => void;
  hooks?: {
    parse?: {
      before?: PrintHookBeforeParse[];
      after?: PrintHookAfterParse[];
    };
    inject?: {
      before?: PrintHookBeforeInject[];
      after?: PrintHookAfterInject[];
    };
    write?: {
      before?: PrintHookBeforeWrite[];
      after?: PrintHookAfterWrite[];
    };
  };
}

function getRouteFileContent(routeFilePath: string): string {
  const stats = fs.statSync(routeFilePath, { throwIfNoEntry: false });
  if (!stats || !stats.isFile()) {
    throw new Error(`Could not find the file with path ${routeFilePath}`);
  }
  return fs.readFileSync(routeFilePath, 'utf-8');
}

const IMPORT_DECLARATIONS_COMMENT = 'injected by sagaroute';
export const IMPORT_DECLARATIONS_COMMENT_START = `${IMPORT_DECLARATIONS_COMMENT}: start`;
export const IMPORT_DECLARATIONS_COMMENT_END = `${IMPORT_DECLARATIONS_COMMENT}: end`;

function removeInjectedContent(content: string) {
  return content.replace(
    /\/\* injected by sagaroute: start \*\/.*\/\* injected by sagaroute: end \*\//gs,
    '',
  );
}

function parseStickyInjectCommentToPlaceholder(
  ast: ParseResult<t.File>,
  handledInjectVariables: string[],
) {
  function getInjectedVariable(node: t.Node) {
    const startLine = node.loc?.start.line;
    const { leadingComments } = node;
    const matchObj = leadingComments
      ?.find(({ loc }) => loc?.start && loc.start.line + 1 === startLine)
      ?.value.match(INJECT_SYMBOL_PREFIX_REGEXP);
    return matchObj?.[1]?.trim();
  }

  traverse(ast, {
    VariableDeclaration({ node }) {
      const startLine = node.loc?.start.line;
      const varName = getInjectedVariable(node);
      if (varName) {
        for (let i = 0; i < node.declarations.length; i++) {
          const declarator = node.declarations[i];
          if (
            t.isVariableDeclarator(declarator) &&
            declarator.loc &&
            /**
             * 防止一条声明语句里被拆分成多行的情况，如：
             * let code=x,
             *  code1=x1
             */
            declarator.loc.start.line === startLine
          ) {
            handledInjectVariables.push(varName);
            declarator.init = t.identifier(`{{{${varName}}}}`);
          }
        }
      }
    },
    ExportNamedDeclaration({ node }) {
      const startLine = node.loc?.start.line;
      const varName = getInjectedVariable(node);
      if (varName && t.isVariableDeclaration(node.declaration)) {
        for (let i = 0; i < node.declaration.declarations.length; i++) {
          const declarator = node.declaration.declarations[i];
          if (
            t.isVariableDeclarator(declarator) &&
            declarator.loc &&
            /**
             * 防止一条声明语句里被拆分成多行的情况，如：
             * let code=x,
             *  code1=x1
             */
            declarator.loc.start.line === startLine
          ) {
            handledInjectVariables.push(varName);
            declarator.init = t.identifier(`{{{${varName}}}}`);
          }
        }
      }
    },
    ReturnStatement({ node }) {
      const varName = getInjectedVariable(node);
      if (varName) {
        handledInjectVariables.push(varName);
        node.argument = t.identifier(`{{{${varName}}}}`);
      }
    },
    ExportDefaultDeclaration({ node }) {
      const varName = getInjectedVariable(node);
      if (varName) {
        handledInjectVariables.push(varName);
        node.declaration = t.identifier(`{{{${varName}}}}`);
      }
    },
  });
}

function parseEmptyInjectCommentToPlaceholder(content: string, handledInjectVariables: string[]) {
  const injectedVariables = new Set(handledInjectVariables);
  function replacer(match: string, p1: string) {
    const matchObj = p1.match(INJECT_SYMBOL_PREFIX_REGEXP);
    if (matchObj) {
      const varName = matchObj[1]?.trim();
      if (!handledInjectVariables.includes(varName)) {
        injectedVariables.add(varName);
        return [
          match,
          `/* ${IMPORT_DECLARATIONS_COMMENT_START} */`,
          `{{{${varName}}}}`,
          `/* ${IMPORT_DECLARATIONS_COMMENT_END} */`,
        ].join('\n');
      }
    }
    return match;
  }

  let modifiedContent = content.replace(/\/\/(.*)/g, replacer);
  modifiedContent = modifiedContent.replace(/\/\*(.*)\*\//g, replacer);
  return {
    template: modifiedContent,
    injectedVariables: Array.from(injectedVariables),
  };
}

function generateRenderTemplateFromContent(content: string, routeFilePath: string) {
  const handledInjectVariables: string[] = [];
  const modifiedContent = removeInjectedContent(content);
  const ast = parseToAst(
    modifiedContent,
    ['.ts', '.tsx'].includes(path.extname(routeFilePath)),
    routeFilePath,
  );
  const existedImportNames = collectExistedImportNames(ast);
  parseStickyInjectCommentToPlaceholder(ast, handledInjectVariables);
  let { code: template } = generate(ast);
  let injectedVariables: string[];
  // eslint-disable-next-line prefer-const
  ({ template, injectedVariables } = parseEmptyInjectCommentToPlaceholder(
    template,
    handledInjectVariables,
  ));
  return { template, existedImportNames, injectedVariables };
}

function collectExistedImportNames(ast: ParseResult<t.File>) {
  const existedImportNames: string[] = [];
  traverse(ast, {
    ImportDeclaration({ node }) {
      node.specifiers.forEach((specifier) => {
        if (
          t.isImportDefaultSpecifier(specifier) ||
          t.isImportNamespaceSpecifier(specifier) ||
          t.isImportSpecifier(specifier)
        ) {
          existedImportNames.push(specifier.local.name);
        }
      });
    },
  });
  return existedImportNames;
}

function transformImportsToString(imports: Imports, existedImportNames: string[]) {
  const importLines: string[] = [];

  Object.entries(imports).forEach(([source, dependencies]) => {
    const notDefaultImport: string[] = [];
    let defaultImport: string = '';
    dependencies
      .filter(({ asName }) => !existedImportNames.includes(asName))
      .forEach((dependency) => {
        if (dependency.isDefault) {
          // import x1 from 'xxx'
          if (dependency.name !== '*') {
            defaultImport = dependency.asName;
            // import * as x1 from 'xxx'
          } else {
            importLines.push(`import * as ${dependency.asName} from "${source}";`);
          }
        } else {
          // import {x1} from 'xxx'
          if (dependency.name === dependency.asName) {
            notDefaultImport.push(dependency.name);
          } else {
            notDefaultImport.push(`${dependency.name} as ${dependency.asName}`);
          }
        }
      });
    const importElementString = [
      defaultImport,
      notDefaultImport.length ? `{${notDefaultImport.join(',')}}` : '',
    ]
      .filter((item) => Boolean(item))
      .join(',');
    if (importElementString) {
      importLines.push(`import ${importElementString} from "${source}";`);
    }
  });
  return importLines.join('\n');
}

function diffVariableBetweenViewAndTemplate(
  viewVariable: string[],
  templateVariable: string[],
  onWarning: NonNullable<PrintOption['onWarning']>,
  routeFilePath: string,
) {
  const variablesOnlyInView = viewVariable.filter((item) => !templateVariable.includes(item));
  const variablesOnlyInTemplate = templateVariable.filter((item) => !viewVariable.includes(item));
  if (variablesOnlyInView.length) {
    onWarning(
      `The RoutingTemplateFile<${routeFilePath}> has no variables such as ${variablesOnlyInView
        .map((item) => `[${item}]`)
        .join(',')} which are needed in view.`,
    );
  }
  if (variablesOnlyInTemplate.length) {
    onWarning(
      `The view of stage<print.inject.before> has no variables such as ${variablesOnlyInTemplate
        .map((item) => `[${item}]`)
        .join(',')} which are needed in template.`,
    );
  }
}

export default function print(routes: RouteObject[], imports: Imports, option?: PrintOption) {
  const { routeFilePath = path.join('src', 'route.ts'), hooks, onWarning } = option ?? {};
  let renderTemplate = hookCompose(hooks?.parse?.before, routeFilePath);
  if (renderTemplate === null) return;
  let existedImportNames: string[] = [];
  let injectedVariables: undefined | string[];
  if (!renderTemplate) {
    const content = getRouteFileContent(routeFilePath);
    ({
      template: renderTemplate,
      existedImportNames,
      injectedVariables,
    } = generateRenderTemplateFromContent(content, routeFilePath));
  }
  hookCompose(hooks?.parse?.after, renderTemplate, routeFilePath);
  const view: Record<string, any> = {
    [TEMPLATE_RENDER_ROUTE_SYMBOL]: transformRoutesToString(routes),
    [TEMPLATE_RENDER_IMPORTS_SYMBOL]: transformImportsToString(imports, existedImportNames),
  };
  let renderedContent = hookCompose(hooks?.inject?.before, view, renderTemplate, routes);
  if (renderedContent === null) return;
  // 如果用户在parse.before中返回了template，此时injectedVariables为undefined
  if (injectedVariables && onWarning) {
    diffVariableBetweenViewAndTemplate(
      Object.keys(view),
      injectedVariables,
      onWarning,
      routeFilePath,
    );
  }
  if (!renderedContent) {
    renderedContent = Mustache.render(renderTemplate, view);
    renderedContent = renderedContent.replace(
      /\/\* injected by sagaroute: start \*\/\s*\/\* injected by sagaroute: end \*\/\s/gs,
      '',
    );
  }
  hookCompose(hooks?.inject?.after, renderedContent, view, renderTemplate);
  if (hookCompose(hooks?.write?.before, renderedContent, routeFilePath) !== null) {
    fs.writeFileSync(routeFilePath, renderedContent, {
      flag: 'w+',
    });
    hookCompose(hooks?.write?.after, routeFilePath);
  }
}
