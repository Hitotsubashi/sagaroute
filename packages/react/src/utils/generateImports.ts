import { Dependency } from '@/gather';
import { Imports } from '@/weave';

export default function generateImports(dependencies: Dependency[] | undefined): Imports {
  const imports: Imports = {};
  dependencies?.forEach((dependency) => {
    const { importPath, ...rest } = dependency;
    if (imports[importPath]) {
      imports[importPath].push(rest);
    } else {
      imports[importPath] = [rest];
    }
  });
  return imports;
}
