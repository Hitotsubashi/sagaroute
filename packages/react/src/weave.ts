import { Dependency, FileNode } from '@/gather';
import path from 'path';
import { EVAL_STRING_SYMBOL } from '@/utils/symbol';
import hookCompose from '@/utils/hookCompose';
import generateLazyImportFn from '@/utils/generateLazyImportFn';
import normalizePath from '@/utils/normalizePath';
import generateImportName from '@/utils/generateImportName';
import mergeImports from '@/utils/mergeImports';
import generateImports from './utils/generateImports';
import { PartialRequired } from './typings';

export type WeaveHookBefore = (fileNodes: FileNode[]) => void | null;
export type WeaveHookBeforeEach = (
  fileNode: FileNode,
  parent: FileNodeParent,
) => void | { route: RouteObject; imports: Imports } | null;
export type WeaveHookAfterEach = (
  route: RouteObject,
  imports: Imports,
  fileNode: FileNode,
  parent: FileNodeParent,
) => void;
export type WeaveHookAfter = (
  routes: RouteObject[],
  imports: Imports,
  fileNodes: FileNode[],
) => void;

interface WeaveOption {
  lazy?: boolean | LazyFn;
  relativeDirpath?: string;
  relativeLayoutDirPath?: string;
  pathRewrite?: [RegExp, string][];
  hooks?: {
    before?: WeaveHookBefore[];
    beforeEach?: WeaveHookBeforeEach[];
    afterEach?: WeaveHookAfterEach[];
    after?: WeaveHookAfter[];
  };
}

type InnerWeaveOption = PartialRequired<WeaveOption, 'relativeDirpath' | 'relativeLayoutDirPath'>;

export type LazyFn = (fpath: string) => boolean;
export interface RouteObject {
  path?: string;
  index?: boolean;
  caseSensitive?: boolean;
  children?: RouteObject[];
  element?: string;
  lazy?: string | LazyFn;
  [key: string]: any;
}

export type Imports = Record<string, Omit<Dependency, 'importPath'>[]>;

export interface FileNodeParent {
  path: string;
  layoutNode?: boolean;
}

function transformParamNameToParamPath(name: string | undefined) {
  if (name === undefined) return name;
  if (name === '[]') return '*';
  let match = name.match(/^\[([^$]*)\$\]$/);
  if (match) {
    return `:${match[1]}?`;
  }
  match = name.match(/^\[([^\]]*)\]$/);
  if (match) {
    return `:${match[1]}`;
  }
  return name;
}

function traverse(
  fileNodes: FileNode[],
  parent: FileNodeParent,
  option: InnerWeaveOption,
): { routes: RouteObject[]; imports: Imports } {
  const { lazy, relativeDirpath, relativeLayoutDirPath, pathRewrite, hooks } = option;
  const imports: Imports = {};
  const routes = fileNodes.reduce((pre, fileNode) => {
    let route: RouteObject | null;
    let routeImports: Imports = {};

    const hookResult = hookCompose(hooks?.beforeEach, fileNode, parent);

    if (!hookResult && hookResult !== null) {
      const { name, children, type, layoutNode, props, dependencies } = fileNode;
      const isFile = type === 'file';
      const currentPath = path.join(parent.path, name);
      // fileNode 类型为文件
      if (isFile) {
        const basename = name.replace(path.extname(name), '');
        let routePath: RouteObject['path'] = layoutNode ? '/' : basename;
        const index =
          props?.routeProps?.index !== undefined
            ? props.routeProps.index
            : !layoutNode && (parent.path !== '' || parent.layoutNode) && routePath === 'index'
            ? true
            : undefined;
        if (index) {
          routePath = undefined;
        } else if (routePath === 'index' && parent.path === '') {
          routePath = '/';
        } else if (routePath === '404') {
          routePath = '*';
        }
        routePath = transformParamNameToParamPath(routePath);
        const childrenResult = layoutNode
          ? traverse(children!, { path: '', layoutNode: true }, option)
          : undefined;
        route = {
          path: routePath,
          index,
          ...props?.routeProps,
          children: childrenResult?.routes,
        };

        const routeImportPath = path.join(
          layoutNode ? relativeLayoutDirPath : relativeDirpath,
          currentPath,
        );
        // routeProps.lazy有定义
        if (props?.routeProps?.lazy) {
          routeImports = generateImports(dependencies);
          // 全局lazy被定义
        } else if (typeof lazy === 'function' ? lazy(fileNode.path) : lazy) {
          const { path, index, children, caseSensitive, ...restRouteProps } = route;
          route = {
            path,
            index,
            children,
            caseSensitive,
            lazy:
              props?.routeProps?.lazy ||
              generateLazyImportFn(routeImportPath, {
                pathRewrite,
                restRouteProps,
                dependencies: fileNode.dependencies,
              }),
          };
          // lazy为false
        } else {
          const asName = generateImportName(routeImportPath, '');
          const importForCurrentRouteComponent: Imports = {
            [normalizePath(routeImportPath, pathRewrite)]: [
              {
                name: basename,
                asName,
                isDefault: true,
              },
            ],
          };
          routeImports = generateImports(dependencies);
          mergeImports(routeImports, importForCurrentRouteComponent);
          route = {
            ...route,
            element: `${EVAL_STRING_SYMBOL}<${asName}/>`,
          };
        }
        if (childrenResult?.imports) {
          mergeImports(routeImports, childrenResult.imports);
        }
        // fileNode 类型为文件夹
      } else {
        const _layoutFileNode = children!.find(
          ({ name, type }) => type === 'file' && name.replace(path.extname(name), '') === '_layout',
        );
        route = {
          path: transformParamNameToParamPath(name),
        };
        // 文件里的文件夹中包含_layout
        if (_layoutFileNode) {
          const childrenResult = traverse(
            children!.filter((item) => item !== _layoutFileNode),
            { path: currentPath },
            option,
          );
          route = {
            ...route,
            children: childrenResult.routes,
            ..._layoutFileNode.props?.routeProps,
          };
          const routeImportPath = path.join(relativeDirpath, currentPath, _layoutFileNode.name);
          // routeProps.lazy有定义
          if (_layoutFileNode.props?.routeProps?.lazy) {
            routeImports = generateImports(dependencies);
            // 全局lazy有定义
          } else if (typeof lazy === 'function' ? lazy(_layoutFileNode.path) : lazy) {
            const { path, index, children, caseSensitive, ...restRouteProps } = route;
            route = {
              path,
              index,
              children,
              caseSensitive,
              lazy:
                _layoutFileNode.props?.routeProps?.lazy ||
                generateLazyImportFn(routeImportPath, {
                  pathRewrite,
                  restRouteProps,
                  dependencies: _layoutFileNode.dependencies,
                }),
            };
            // lazy为false
          } else {
            const asName = generateImportName(routeImportPath, '');
            const basename = _layoutFileNode.name.replace(path.extname(_layoutFileNode.name), '');
            const importForCurrentRouteComponent: Imports = {
              [normalizePath(routeImportPath, pathRewrite)]: [
                {
                  name: basename,
                  asName,
                  isDefault: true,
                },
              ],
            };
            routeImports = generateImports(dependencies);
            mergeImports(routeImports, importForCurrentRouteComponent);
            route = {
              ...route,
              element: `${EVAL_STRING_SYMBOL}<${asName}/>`,
            };
          }
          if (childrenResult.imports) {
            mergeImports(routeImports, childrenResult.imports);
          }
          // 文件里的文件夹中不包含_layout
        } else {
          const childrenResult = traverse(children!, { path: currentPath }, option);
          route = {
            ...route,
            children: childrenResult.routes,
          };
          if (childrenResult.imports) {
            mergeImports(routeImports, childrenResult.imports);
          }
        }
      }
      route = normalize(route);
    } else if (hookResult === null) {
      route = null;
    } else {
      ({ route, imports: routeImports } = hookResult);
    }
    if (route === null) {
      return pre;
    }
    if (route) {
      hookCompose(hooks?.afterEach, route, routeImports, fileNode, parent);
      mergeImports(imports, routeImports);
      return pre.concat(route);
    }
    return pre;
  }, [] as RouteObject[]);
  return { routes, imports };
}

function normalize(route: RouteObject): RouteObject {
  for (const key in route) {
    if (route[key] === undefined) {
      delete route[key];
    }
  }
  if (Array.isArray(route.children)) {
    if (!route.children.length) {
      delete route.children;
    }
  }
  return route;
}

export default function weave(fileNodes: FileNode[], option: WeaveOption) {
  const {
    relativeDirpath = path.join('..', 'src', 'pages'),
    relativeLayoutDirPath = path.join('..', 'src', 'layouts'),
  } = option;
  if (hookCompose(option.hooks?.before, fileNodes) !== null) {
    const layoutFileNode = fileNodes.find(({ layoutNode }) => layoutNode);
    let ultimateFileNodes: FileNode[] = fileNodes;
    if (layoutFileNode) {
      const unlayoutFileNodes = fileNodes.filter(
        ({ name, props, layoutNode, children }) =>
          !layoutNode &&
          ((name.replace(path.extname(name), '') === '404' &&
            props?.routeOptions?.layout !== true) ||
            props?.routeOptions?.layout === false ||
            children?.some(
              ({ type, name, props }) =>
                type === 'file' &&
                name.replace(path.extname(name), '') === '_layout' &&
                props?.routeOptions?.layout === false,
            )),
      );

      ultimateFileNodes = [
        {
          ...layoutFileNode,
          children: fileNodes.filter(
            (item) => item !== layoutFileNode && !unlayoutFileNodes.includes(item),
          ),
        },
        ...unlayoutFileNodes,
      ];
    }
    const { routes, imports } = traverse(
      ultimateFileNodes,
      { path: '' },
      { ...option, relativeDirpath, relativeLayoutDirPath },
    );
    hookCompose(option?.hooks?.after, routes, imports, fileNodes);
    return { routes, imports };
  } else {
    return null;
  }
}
