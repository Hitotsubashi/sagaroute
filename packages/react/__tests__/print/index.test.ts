import fs from "fs";
import path from "path";
import { EVAL_STRING_SYMBOL } from "@/utils/symbol";
import print from "@/print";
import { Imports, RouteObject } from "@/weave";

test("test print with normal routes by comment: //", () => {
  const routes: RouteObject[] = [
    {
      path: "users",
      children: [
        {
          path: ":id",
          element: `${EVAL_STRING_SYMBOL}<SrcPagesUsers_Id_/>`,
        },
      ],
    },
    {
      path: "/",
      element: `${EVAL_STRING_SYMBOL}<SrcPagesIndex/>`,
    },
    {
      path: "*",
      element: `${EVAL_STRING_SYMBOL}<SrcPages404/>`,
    },
  ];
  const imports: Imports = {
    "./src/pages/users/[id].tsx": [
      {
        name: "User",
        asName: "SrcPagesUsers_Id_",
        isDefault: true,
      },
    ],
    "./src/pages/index.tsx": [
      {
        name: "Index",
        asName: "SrcPagesIndex",
        isDefault: true,
      },
    ],
    "./src/pages/404.tsx": [
      {
        name: "Page404",
        asName: "SrcPages404",
        isDefault: true,
      },
    ],
  };
  const routeFilePath = path.join(
    __dirname,
    "ignore-test-render",
    "route1.tsx"
  );
  print(routes, imports, { routeFilePath });
  const context = fs.readFileSync(routeFilePath, "utf-8");
  expect(context).toMatchSnapshot();
});

test("test print with normal routes by comment: /**/", () => {
  const routes: RouteObject[] = [
    {
      path: "users",
      children: [
        {
          path: ":id",
          element: `${EVAL_STRING_SYMBOL}<SrcPagesUsers_Id_/>`,
        },
      ],
    },
    {
      path: "/",
      element: `${EVAL_STRING_SYMBOL}<SrcPagesIndex/>`,
    },
    {
      path: "*",
      element: `${EVAL_STRING_SYMBOL}<SrcPages404/>`,
    },
  ];
  const imports: Imports = {
    "./src/pages/users/[id].tsx": [
      {
        name: "User",
        asName: "SrcPagesUsers_Id_",
        isDefault: true,
      },
    ],
    "./src/pages/index.tsx": [
      {
        name: "Index",
        asName: "SrcPagesIndex",
        isDefault: true,
      },
    ],
    "./src/pages/404.tsx": [
      {
        name: "Page404",
        asName: "SrcPages404",
        isDefault: true,
      },
    ],
  };
  const routeFilePath = path.join(
    __dirname,
    "ignore-test-render",
    "route2.tsx"
  );
  print(routes, imports, { routeFilePath });
  const context = fs.readFileSync(routeFilePath, "utf-8");
  expect(context).toMatchSnapshot();
});

test("print with multiple imports", () => {
  const routes: RouteObject[] = [
    {
      path: "users",
      element: `${EVAL_STRING_SYMBOL}<PagesUsers/>`,
      action: `${EVAL_STRING_SYMBOL}action1`,
      loader1: `${EVAL_STRING_SYMBOL}loader.loader1`,
    },
    {
      path: "dashboard",
      Component: `${EVAL_STRING_SYMBOL}<PagesDashboard/>`,
      action: `${EVAL_STRING_SYMBOL}action2`,
      loader1: `${EVAL_STRING_SYMBOL}loader.loader2`,
    },
    {
      path: "chart",
      Component: `${EVAL_STRING_SYMBOL}<PagesChart/>`,
      action: `${EVAL_STRING_SYMBOL}defaultAction`,
      loader1: `${EVAL_STRING_SYMBOL}loader3`,
      errorElement: `${EVAL_STRING_SYMBOL}<PagesChartError/>`,
    },
  ];

  const imports: Imports = {
    "@/utils/action.ts": [
      {
        name: "action1",
        asName: "action1",
        isDefault: false,
      },
      {
        name: "action2",
        asName: "action2",
        isDefault: false,
      },
      {
        name: "default",
        asName: "defaultAction",
        isDefault: false,
      },
    ],
    "@/utils/loader.ts": [
      {
        name: "*",
        asName: "loader",
        isDefault: true,
      },
      {
        name: "loader3",
        asName: "loader3",
        isDefault: false,
      },
    ],
    "@/pages/users.tsx": [
      {
        name: "users",
        asName: "PagesUsers",
        isDefault: true,
      },
    ],
    "@/pages/dashboard.tsx": [
      {
        name: "dashboard",
        asName: "PagesDashboard",
        isDefault: true,
      },
    ],
    "@/pages/chart.tsx": [
      {
        name: "chart",
        asName: "PagesChart",
        isDefault: true,
      },
      {
        name: "Error",
        asName: "PagesChartError",
        isDefault: false,
      },
    ],
  };

  const routeFilePath = path.join(
    __dirname,
    "ignore-test-render",
    "route8.tsx"
  );
  print(routes, imports, { routeFilePath });
  const context = fs.readFileSync(routeFilePath, "utf-8");
  expect(context).toMatchSnapshot();
});
