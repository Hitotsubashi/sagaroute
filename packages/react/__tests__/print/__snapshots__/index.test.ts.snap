// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`print with multiple imports 1`] = `
"// @ts-nocheck
/* sagaroute-inject:imports */
/* injected by sagaroute: start */
import {action1,action2,default as defaultAction} from "@/utils/action.ts";
import * as loader from "@/utils/loader.ts";
import {loader3} from "@/utils/loader.ts";
import PagesUsers from "@/pages/users.tsx";
import PagesDashboard from "@/pages/dashboard.tsx";
import PagesChart,{Error as PagesChartError} from "@/pages/chart.tsx";
/* injected by sagaroute: end */

import React from "react";

// sagaroute-inject: routes
export default [
  {
    "path": "users",
    "element": <PagesUsers/>,
    "action": action1,
    "loader1": loader.loader1
  },
  {
    "path": "dashboard",
    "Component": <PagesDashboard/>,
    "action": action2,
    "loader1": loader.loader2
  },
  {
    "path": "chart",
    "Component": <PagesChart/>,
    "action": defaultAction,
    "loader1": loader3,
    "errorElement": <PagesChartError/>
  }
];"
`;

exports[`test print with normal routes by comment: /**/ 1`] = `
"// @ts-nocheck
import React from "react";
/* sagaroute-inject:imports */
/* injected by sagaroute: start */
import SrcPagesUsers_Id_ from "./src/pages/users/[id].tsx";
import SrcPagesIndex from "./src/pages/index.tsx";
import SrcPages404 from "./src/pages/404.tsx";
/* injected by sagaroute: end */

function getRoutes() {
  /* sagaroute-inject:routes */
  return [
  {
    "path": "users",
    "children": [
      {
        "path": ":id",
        "element": <SrcPagesUsers_Id_/>
      }
    ]
  },
  {
    "path": "/",
    "element": <SrcPagesIndex/>
  },
  {
    "path": "*",
    "element": <SrcPages404/>
  }
];
}
export default getRoutes;"
`;

exports[`test print with normal routes by comment: // 1`] = `
"// @ts-nocheck
import React from "react";
// @ts-ignore
/* sagaroute-inject:imports */
/* injected by sagaroute: start */
import SrcPagesUsers_Id_ from "./src/pages/users/[id].tsx";
import SrcPagesIndex from "./src/pages/index.tsx";
import SrcPages404 from "./src/pages/404.tsx";
/* injected by sagaroute: end */

// sagaroute-inject: routes
const routes = [
  {
    "path": "users",
    "children": [
      {
        "path": ":id",
        "element": <SrcPagesUsers_Id_/>
      }
    ]
  },
  {
    "path": "/",
    "element": <SrcPagesIndex/>
  },
  {
    "path": "*",
    "element": <SrcPages404/>
  }
];
export default routes;"
`;
